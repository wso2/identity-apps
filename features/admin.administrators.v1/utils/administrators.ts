/**
 * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import {
    UserBasicInterface,
    UserRoleInterface
} from "@wso2is/admin.core.v1";
import { administratorConfig } from "@wso2is/admin.extensions.v1/configs/administrator";
import { SCIMConfigs } from "@wso2is/admin.extensions.v1/configs/scim";
import { UserAccountTypes } from "@wso2is/admin.users.v1/constants/user-management-constants";

/**
 * Checks whether administrator role is present in the user.
 * @param user - user to check if the user is an admin
 * @returns boolean - true if the user is an admin.
 */
export const isAdminUser = (user: UserBasicInterface): boolean => {
    if (!user?.roles) {
        return false;
    }

    return user.roles.some((role: UserRoleInterface) =>
        role.display === administratorConfig.adminRoleName
    );
};

/**
 * Checks whether the user is the owner of the tenant.
 * @param user - user to check if the user is the owner
 * @returns boolean - true if the user is the owner.
 */
export const isOwner = (user: UserBasicInterface):boolean =>
    user[ SCIMConfigs.scim.enterpriseSchema ]?.userAccountType === UserAccountTypes.OWNER;

/**
 * Checks whether the user is a collaborator.
 * Collaborator users does not have a username with a domain separeated by a `/`.
 * primary/username is NOT a collaborator user.
 *
 * @param user - user to check if the user is a collaborator
 * @returns boolean - true if the user is a collaborator.
 */
export const isCollaboratorUser = (user: UserBasicInterface): boolean =>
    user?.userName?.split("/")?.length === 1;
