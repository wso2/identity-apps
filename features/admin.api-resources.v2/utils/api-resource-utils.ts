/**
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { FeatureConfigInterface } from "@wso2is/admin.core.v1";
import { hasRequiredScopes } from "@wso2is/core/helpers";
import { APIResourceCategories, APIResourcesConstants } from "../constants/api-resources-constants";
import { Property } from "../models/api-resources";

export class APIResourceUtils {

    /**
     * Private constructor to avoid object instantiation from outside
     * the class.
     */
    private constructor() { }

    /**
     * Check whether the API resource read is allowed.
     *
     * @param featureConfig - Feature config.
     * @param allowedScopes - Allowed scopes.
     * @returns True if the API resource read is allowed.
     */
    public static isAPIResourceReadAllowed(featureConfig: FeatureConfigInterface,
        allowedScopes: string): boolean {

        return hasRequiredScopes(featureConfig?.apiResources,
            featureConfig?.apiResources?.scopes?.read, allowedScopes);
    }

    /**
     * Check whether the API resource update is allowed.
     *
     * @param featureConfig - Feature config.
     * @param allowedScopes - Allowed scopes.
     * @returns True if the API resource update is allowed.
     */
    public static isAPIResourceUpdateAllowed(featureConfig: FeatureConfigInterface,
        allowedScopes: string): boolean {

        return hasRequiredScopes(featureConfig?.apiResources,
            featureConfig?.apiResources?.scopes?.update, allowedScopes);
    }

    /**
     * Check whether the API resource create is allowed.
     *
     * @param featureConfig - Feature config.
     * @param allowedScopes - Allowed scopes.
     * @returns True if the API resource create is allowed.
     */
    public static isAPIResourceCreateAllowed(featureConfig: FeatureConfigInterface,
        allowedScopes: string): boolean {

        return hasRequiredScopes(featureConfig?.apiResources,
            featureConfig?.apiResources?.scopes?.create, allowedScopes);
    }

    /**
     * Check whether the API resource delete is allowed.
     *
     * @param featureConfig - Feature config.
     * @param allowedScopes - Allowed scopes.
     * @returns True if the API resource delete is allowed.
     */
    public static isAPIResourceDeleteAllowed(featureConfig: FeatureConfigInterface,
        allowedScopes: string): boolean {

        return hasRequiredScopes(featureConfig?.apiResources,
            featureConfig?.apiResources?.scopes?.delete, allowedScopes);
    }

    /**
     * Check whether the API resource is a system API.
     *
     * @param type - API Resource type.
     * @returns True if the API resource is a system API.
     */
    public static isSystemAPI(type: string): boolean {

        return type !== APIResourcesConstants.BUSINESS;
    }

    public static resolveApiResourceGroup = (groupName: string): string => {
        switch (groupName) {
            case APIResourceCategories.TENANT:
                return "Management APIs";
            case APIResourceCategories.ORGANIZATION:
                return "Organization APIs";
            case APIResourceCategories.BUSINESS:
                return "Business APIs";
            default:
                return groupName;
        }
    };

    public static resolveApiResourceGroupDisplayName = (groupName: string): string => {
        switch (groupName) {
            case APIResourceCategories.ORGANIZATION:
                return "Organization API";
            case APIResourceCategories.TENANT:
                return "Management API";
            case APIResourceCategories.CONSOLE_FEATURE:
                return "Console Feature";
            case APIResourceCategories.BUSINESS:
                return "Business API";
            default:
                return groupName;
        }
    };

    public static resolveApiResourceGroupDescription = (groupName: string): string => {
        switch (groupName) {
            case APIResourceCategories.ORGANIZATION:
                return "extensions:develop.apiResource.organizationAPI.description";
            case APIResourceCategories.TENANT:
                return "extensions:develop.apiResource.managementAPI.description";
            case APIResourceCategories.CONSOLE_FEATURE:
                return "extensions:develop.apiResource.consoleFeature.description";
            case APIResourceCategories.BUSINESS:
                return "extensions:develop.apiResource.businessAPI.description";
            default:
                return groupName;
        }
    };

    /**
     * Checks if the API resource is managed by Choreo.
     *
     * @param properties - API resource properties.
     * @returns - `true` if the API resource is managed by Choreo and `false` otherwise.
     */
    public static checkIfAPIResourcePropertyManaged = (properties: Property[]): boolean => {
        if (properties && properties?.length > 0) {
            const isChoreoAPIProperty: Property =
            properties.find((property: Property) => property?.name === APIResourcesConstants.IS_CHOREO_API);

            return isChoreoAPIProperty ? isChoreoAPIProperty?.value === "true" : false;
        }

        return false;
    }
}
