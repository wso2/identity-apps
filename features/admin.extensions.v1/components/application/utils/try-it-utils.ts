/**
 * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { ApplicationManagementConstants } from "@wso2is/admin.applications.v1/constants";
import { AppState, AppUtils } from "@wso2is/admin.core.v1";
import { StorageIdentityAppsSettingsInterface } from "@wso2is/core/models";
import cloneDeep from "lodash-es/cloneDeep";
import get from "lodash-es/get";
import isEmpty from "lodash-es/isEmpty";
import set from "lodash-es/set";
import { FunctionComponent } from "react";
import { useSelector } from "react-redux";
import LoginApplicationTemplate from
    "../../../application-templates/templates/single-page-application/login-playground-application.json";
import { TryItApplicationConstants } from "../constants/try-it-constants";

/**
 * Persist the Playground Tour seen status in Local Storage.
 *
 * @param {boolean} status - Status to set.
 */
export const persistPlaygroundTourViewedStatus = (status: boolean): void => {
    const userPreferences: StorageIdentityAppsSettingsInterface = AppUtils.getUserPreferences();
    const newPref: StorageIdentityAppsSettingsInterface = cloneDeep(userPreferences);

    set(newPref?.identityAppsSettings?.devPortal,
        TryItApplicationConstants.TRY_IT_TOUR_STATUS_STORAGE_KEY, status);

    AppUtils.setUserPreferences(newPref);
};

/**
 * Check if the Playground Tour has already been seen by the user.
 *
 * @return {boolean}
 */
export const getPlaygroundTourViewedStatus = (): boolean => {
    const userPreferences: StorageIdentityAppsSettingsInterface = AppUtils.getUserPreferences();

    if (isEmpty(userPreferences)) {
        return false;
    }

    return get(userPreferences?.identityAppsSettings?.devPortal,
        TryItApplicationConstants.TRY_IT_TOUR_STATUS_STORAGE_KEY, false);
};

/**
 * Get the client Id of the Try it application.
 *
 * @return {string}
 */
export const getTryItClientId = (tenantDomain:string): string => {

    return LoginApplicationTemplate.application.
        inboundProtocolConfiguration.oidc.clientId.replace("<TENANT>", tenantDomain);
};
