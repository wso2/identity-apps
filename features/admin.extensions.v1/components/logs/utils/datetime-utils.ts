/**
 * Copyright (c) 2022-2024, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import moment from "moment";

/**
 * Returns a date-time string in human readable format
 *
 * @param dateString - (string)
 * @returns formatted date time string i.e., "16 Sep, 2022 | 15:19:16"
 */
const formatTimestampToDateTime = (dateString: string): string => {
    if (!dateString) return "N/A";

    return moment(dateString).format("DD MMM, YYYY | kk:mm:ss");
};

/**
 * Returns a date string in human readable format
 *
 * @param dateString - (string)
 * @returns formatted date time string i.e., "16 Sep, 2022"
 */
const getDateFromTimestamp = (dateString: string): string => {
    if (!dateString) return "N/A";

    return moment(dateString).format("DD MMM, YYYY");
};

/**
 * Returns a time string in human readable format
 *
 * @param dateString - (string)
 * @returns formatted date time string i.e., "15:19:16"
 */
const getTimeFromTimestamp = (dateString: string): string => {
    if (!dateString) return "N/A";

    return moment(dateString).format("kk:mm:ss");
};


/**
 * Returns converted date-time UNIX epoch value using time zone offset
 */
const getDateTimeWithOffset = (timeZone: string): number => {
    const currentLocalTime: Date = new Date();
    const convertedTime: string = `${currentLocalTime.toLocaleString("en-US", { timeZone: timeZone.split(" ")[1] })}` +
        ` ${timeZone.split(" ")[0]}`;
    const UTCValue: number = Date.parse(convertedTime);

    return UTCValue;
};

/**
 * Get the current time zone using local time.
 */
const getCurrentTimeZone = (): string => {
    // Create a new Date object
    const date: Date = new Date();

    // Get the timezone offset in minutes
    const offset: number = date.getTimezoneOffset();

    // Convert the offset to hours and minutes
    const absOffset: number = Math.abs(offset);
    const hours: number = Math.floor(absOffset / 60);
    const minutes: number = absOffset % 60;

    // Determine if the offset is positive or negative
    const sign: string = offset <= 0 ? "+" : "-";

    // Format hours and minutes to always have two digits
    const formattedHours: string = String(hours).padStart(2, "0");
    const formattedMinutes: string = String(minutes).padStart(2, "0");

    // Return the formatted timezone offset
    return `${sign}${formattedHours}${formattedMinutes}`;
};

/**
 * Resolve the maximum time that can be set for the From param based on the given conditions.
 *
 * @param startDate - The start date in YYYY-MM-DD format.
 * @param endDate - The end date in YYYY-MM-DD format.
 * @param timezoneOffset - The timezone offset in the format +HHMM or -HHMM.
 * @param endTime - The end time as a string in HH:mm format, or null if not defined.
 * @returns The maximum from time as a string in HH:mm format.
 */
const resolveMaxFromTime = (
    startDate: string,
    endDate: string,
    timezoneOffset: string,
    endTime: string | null
): string => {
    const date: Date = new Date();

    // Parse the timezone offset
    const sign: string = timezoneOffset[0];
    const hoursOffset: number = parseInt(timezoneOffset.slice(1, 3), 10);
    const minutesOffset: number = parseInt(timezoneOffset.slice(3, 5), 10);
    const totalOffsetMinutes: number = (hoursOffset * 60 + minutesOffset) * (sign === "+" ? 1 : -1);

    // Get the current time in UTC
    const utcHours: number = date.getUTCHours();
    const utcMinutes: number = date.getUTCMinutes();

    // Calculate the local time in the specified timezone
    const localMinutes: number = utcMinutes + totalOffsetMinutes;
    const localHours: number = utcHours + Math.floor(localMinutes / 60);
    const localMinutesRemainder: number = localMinutes % 60;

    // Adjust hours and minutes to fit into 24-hour format
    const adjustedHours: number = (localHours + 24) % 24;
    const adjustedMinutes: number = (localMinutesRemainder + 60) % 60;

    // Format the time as HH:mm
    const formattedHours: string = adjustedHours.toString().padStart(2, "0");
    const formattedMinutes: string = adjustedMinutes.toString().padStart(2, "0");
    const currentTime: string = `${formattedHours}:${formattedMinutes}`;

    const today: string = date.toISOString().split("T")[0];

    // If from date is current date OR if its same as to date,
    // it should be less than the end time IF end time is defined.
    // Otherwise it should be less than the current time based on the timezone
    if (startDate === endDate || startDate === today) {

        return endTime ?? currentTime;
    }

    // Default case: no restriction
    return "23:59";
};

/**
 * Resolve the maximum time that can be set for the To param based on the given conditions.
 *
 * @param startDate - The start date in YYYY-MM-DD format.
 * @param endDate - The end date in YYYY-MM-DD format.
 * @param timezoneOffset - The timezone offset in the format +HHMM or -HHMM.
 * @param startTime - The start time as a string in HH:mm format.
 * @returns The maximum to time as a string in HH:mm format.
 */
const resolveMaxToTime = (endDate: string, timezoneOffset: string): string => {
    const date: Date= new Date();

    // Parse the timezone offset
    const sign: string = timezoneOffset[0];
    const hoursOffset: number = parseInt(timezoneOffset.slice(1, 3), 10);
    const minutesOffset: number = parseInt(timezoneOffset.slice(3, 5), 10);
    const totalOffsetMinutes: number = (hoursOffset * 60 + minutesOffset) * (sign === "+" ? 1 : -1);

    // Get the current time in UTC
    const utcHours: number = date.getUTCHours();
    const utcMinutes: number = date.getUTCMinutes();

    // Calculate the local time in the specified timezone
    const localMinutes: number = utcMinutes + totalOffsetMinutes;
    const localHours: number = utcHours + Math.floor(localMinutes / 60);
    const localMinutesRemainder: number = localMinutes % 60;

    // Adjust hours and minutes to fit into 24-hour format
    const adjustedHours: number = (localHours + 24) % 24;
    const adjustedMinutes: number = (localMinutesRemainder + 60) % 60;

    // Format the time as HH:mm
    const formattedHours: string = adjustedHours.toString().padStart(2, "0");
    const formattedMinutes: string = adjustedMinutes.toString().padStart(2, "0");
    const currentTime: string = `${formattedHours}:${formattedMinutes}`;

    const today: string = date.toISOString().split("T")[0];

    // If to date is current date, it should be less than the current time based on the timezone
    if (endDate === today) {
        return currentTime;
    }

    // Default case: no restriction
    return "23:59";
};

export {
    formatTimestampToDateTime,
    getCurrentTimeZone,
    getDateFromTimestamp,
    getTimeFromTimestamp,
    getDateTimeWithOffset,
    resolveMaxFromTime,
    resolveMaxToTime
};
