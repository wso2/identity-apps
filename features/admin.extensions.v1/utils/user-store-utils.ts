/**
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { getAUserStore } from "@wso2is/admin.core.v1/api";
import { SharedUserStoreConstants } from "@wso2is/admin.core.v1/constants";
import { getUserStoreList } from "@wso2is/admin.userstores.v1/api";
import { UserStoreListItem, UserStorePostData, UserStoreProperty } from "@wso2is/admin.userstores.v1/models";
import { AxiosResponse } from "axios";

/**
 * Utility class for common user store operations.
 */
export class UserStoreUtils {

    private static primaryUserStore:string = "PRIMARY";

    /**
     * Private constructor to avoid object instantiation from outside
     * the class.
     *
     */
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    private constructor() { }

    /**
     * The following method fetch the user store ids list.
     *
     * @param userstores - Externally provided usertores list.
     * @returns userstores list
     */
    public static async getUserStoreIds(userstores?: UserStoreListItem[]): Promise<string[] | void> {

        const getIds = (userstoresList: UserStoreListItem[]): string[] => {
            const userStoreIds: string[] = userstoresList.map((userStore: UserStoreListItem) => userStore.id);

            return userStoreIds;
        };

        if (userstores) {
            return getIds(userstores);
        }

        return getUserStoreList()
            .then((response: AxiosResponse<UserStoreListItem[]>) => {
                return getIds(response.data);
            })
            .catch(() => {
                // Add debug logs here one a logger is added.
                // Tracked here https://github.com/wso2/product-is/issues/11650.
            });
    }

    /**
     * The following method fetch the readonly user stores list.
     *
     * @param userstores - Externally provided usertores list.
     */
    public static async getReadOnlyUserStores(userstores?: UserStoreListItem[]): Promise<string[]> {
        const ids: string[] = await UserStoreUtils.getUserStoreIds(userstores) as string[];
        const readOnlyUserStores: string[] = [];

        readOnlyUserStores.push(UserStoreUtils.primaryUserStore);

        ids.forEach((id: string) => {
            getAUserStore(id)
                .then((res: UserStorePostData) => {
                    res.properties.forEach((property: UserStoreProperty) => {
                        if (property.name === SharedUserStoreConstants.READONLY_USER_STORE
                            && property.value === "true") {
                            readOnlyUserStores.push(res.name.toUpperCase());
                        }
                    });
                }).catch(() => {
                    // Add debug logs here one a logger is added.
                    // Tracked here https://github.com/wso2/product-is/issues/11650.
                });
        });

        return readOnlyUserStores;
    }
}
