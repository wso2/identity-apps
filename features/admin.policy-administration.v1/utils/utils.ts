/**
 * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export const formatXML = (xml: string): string => {
    if (!xml || typeof xml !== "string") {
        console.warn("Invalid XML input for formatting.");
        return xml || "";
    }

    const PADDING: string = " ".repeat(4); // Indentation level
    const regex: RegExp = /(>)(<)(\/*)/g;

    try {
        // Add line breaks between tags
        const formatted = xml.replace(regex, "$1\n$2$3");

        // Detect if already formatted (basic check for line breaks and indentation)
        const alreadyFormatted = formatted.split("\n").some((line) => line.startsWith(PADDING));

        if (alreadyFormatted) {
            return xml; // Return unmodified if already formatted
        }

        let pad = 0;

        return formatted
            .split("\n")
            .map((line: string) => {
                if (line.match(/<\/\w/)) pad -= 1;

                const indent = PADDING.repeat(Math.max(pad, 0));
                const formattedLine = `${indent}${line}`;

                if (line.match(/<\w[^>]*[^/]>.*$/)) pad += 1;

                return formattedLine;
            })
            .join("\n");
    } catch (error) {
        console.error("Error formatting XML:", error);
        return xml; // Return raw XML in case of errors
    }
};

export const unformatXML = (xml: string): string => {
    if (!xml || typeof xml !== "string") {
        console.warn("Invalid XML input for unformatting.");
        return xml || "";
    }

    try {
        // Remove all line breaks, tabs, and excessive spaces
        return xml
            .replace(/>\s+</g, "><") // Remove spaces between tags
            .replace(/\s{2,}/g, " ") // Replace multiple spaces with a single space
            .trim(); // Remove leading/trailing spaces
    } catch (error) {
        console.error("Error unformatting XML:", error);
        return xml; // Return the original XML if unformatting fails
    }
};

