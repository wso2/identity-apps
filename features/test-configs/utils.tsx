/**
 * Copyright (c) 2021-2024, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// import { AuthProvider } from "@asgardeo/auth-react";
import {  RenderResult, render as rtlRender } from "@testing-library/react";
import { AccessControlProvider } from "@wso2is/access-control";
import React, { PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { mockStore } from "./__mocks__/redux/redux-store";
import ReduxStoreStateMock from "./__mocks__/redux/redux-store-state";
// import { AuthenticateUtils } from "../src/features/authentication/utils/authenticate-utils";
// import { PreLoader } from "../src/features/core/components/pre-loader/pre-loader";

/**
 * Custom render method to includes things like global context providers, data stores, etc.
 * @see {@link https://testing-library.com/docs/react-testing-library/setup#custom-render} for more info.
 *
 * @param ui - Component to render.
 * @param allowedScopes - Set of allowed scopes for the logged in user.
 * @param featureConfig - UI Features configuration i.e permissions etc.
 * @param initialState - Redux store initial state.
 * @param store - Mocked store.
 * @param renderOptions - Render options.
 *
 * @returns RenderResult
 */
const render = (
    ui: ReactElement,
    {
        allowedScopes = "internal_login",
        featureConfig = window[ "AppUtils" ].getConfig().ui.features,
        initialState = ReduxStoreStateMock,
        store = mockStore(initialState),
        ...renderOptions
    }: any = {}
): RenderResult => {

    const Wrapper = (props: PropsWithChildren): ReactElement => {

        const { children } = props;

        return (
        // <AuthProvider
        //     config={ AuthenticateUtils.getInitializeConfig() }
        //     fallback={ <PreLoader /> }
        //     getAuthParams={ AuthenticateUtils.getAuthParams }
        // >
            <Provider store={ store }>
                <AccessControlProvider
                    allowedScopes={ allowedScopes }
                    features={ featureConfig }
                    organizationType="FIRST_LEVEL_ORGANIZATION"
                >
                    { children }
                </AccessControlProvider>
            </Provider>
            // </AuthProvider>
        );
    };

    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
