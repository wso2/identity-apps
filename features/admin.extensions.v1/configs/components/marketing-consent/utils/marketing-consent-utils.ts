/**
 * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { LocalStorageUtils } from "@wso2is/core/utils";
import get from "lodash-es/get";
import { MarketingConsentConstants } from "../constants";

/**
 * Util function to get whether the user has interacted with 
 * the marketing consent modal or not from the local storage.
 * 
 * @param uuid - user id.
 * @returns whether the user has interacted with the marketing consent modal or not.
 */
export const getMarketingConsentStatusFromLocalStorage = (uuid: string): boolean => {
    const marketingConsentData: any = JSON.parse(LocalStorageUtils
        .getValueFromLocalStorage(MarketingConsentConstants.LOCAL_STORAGE_KEY));
    const isMarketingConsentGiven: boolean = get(marketingConsentData, uuid, false);

    return isMarketingConsentGiven;
};

/**
 * Util function to store in the local storage that the user has interacted with 
 * the marketing consent modal.
 * 
 * @param uuid - user id.
 */
export const setMarketingConsentStatusToLocalStorage = (uuid: string): void => {
    LocalStorageUtils.setValueInLocalStorage(MarketingConsentConstants.LOCAL_STORAGE_KEY, JSON.stringify({
        ...JSON.parse(LocalStorageUtils.getValueFromLocalStorage(MarketingConsentConstants.LOCAL_STORAGE_KEY)),
        [ uuid ]: true
    }));
};
