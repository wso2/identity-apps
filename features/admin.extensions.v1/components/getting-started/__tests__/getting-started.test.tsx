/**
 * Copyright (c) 2020, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { render, screen } from "@wso2is/unit-testing/utils";
import React from "react";
import { fullPermissions } from "./__mocks__/getting-started-permissions";
import GettingStartedPage from "../getting-started";
import "@testing-library/jest-dom";

describe("Test if the Getting Started page is working as expected", () => {

    // Skipping until the snapshot issue is sorted out.
    it.skip("<GettingStartedPage /> matches snapshot", () => {
        const { container } = render(<GettingStartedPage />, {
            allowedScopes: fullPermissions
        });

        expect(container).toMatchSnapshot();
    });

    it.skip("<GettingStartedPage /> renders without exploding", () => {
        render(<GettingStartedPage />, {
            allowedScopes: fullPermissions
        });

        expect(screen.getByTestId("getting-started-page-layout")).toBeInTheDocument();
    });

    it.skip("<GettingStartedPage /> renders loaders properly", async () => {
        render(<GettingStartedPage />, {
            allowedScopes: fullPermissions
        });

        expect(screen.getByTestId(
            "dynamic-application-context-card-navigate-to-application-list-button"
        )).toBeInTheDocument();

        expect(screen.getByTestId("getting-started-page-layout")).toBeInTheDocument();
        expect(screen.getByTestId("getting-started-page-add-user-button")).toHaveClass("ui loading");
        expect(screen.getByTestId("develop-getting-started-page-add-social-login")).toBeInTheDocument();
        expect(screen.getByTestId("develop-getting-started-page-view-docs")).toBeInTheDocument();
    });
});
