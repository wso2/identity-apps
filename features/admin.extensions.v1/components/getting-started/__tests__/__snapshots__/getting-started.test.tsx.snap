// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if the Getting Started page is working as expected <GettingStartedPage /> matches snapshot 1`] = `
<div>
  <div
    class="layout page-layout getting-started-page"
    data-componentid="getting-started-page-layout"
    data-testid="page-layout"
  >
    <div
      class="layout-content"
    >
      <div
        class="advance-user-view-cards-wrapper"
      >
        <div
          class="greeting"
        >
          <h1
            class="ui header heading bold-500 compact ellipsis"
            data-componentid="welcome-greeting-header"
            data-suppress=""
            data-testid="welcome-greeting-header"
          >
            <div
              class="ui placeholder"
              data-componentid="welcome-greeting-placeholder"
              data-testid="welcome-greeting-placeholder"
            >
              <div
                class="header"
              >
                <div
                  class="line very long"
                />
              </div>
            </div>
          </h1>
        </div>
        <div
          class="ui stackable grid"
        >
          <div
            class="two column row"
          >
            <div
              class="stretched ten wide computer ten wide large screen sixteen wide mobile sixteen wide tablet ten wide widescreen column"
            >
              <div
                class="ui stackable grid"
              >
                <div
                  class="two column row"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
